src/scheduler/DistributedCoordination.cpp

```diff
@@ -0,0 +1,236 @@
+{
+    appSizes.clear();
+
+    barriers.clear();
+
+    recursiveMutexes.clear();
+    mutexes.clear();
+
+    counts.clear();
+    cvs.clear();
+}
+
+void DistributedCoordination::localLockRecursive(int32_t appId)
+{
+    FROM_MAP(mx, std::recursive_mutex, recursiveMutexes);
+    mx->lock();
+}
+
+void DistributedCoordination::localLock(int32_t appId)
+{
+    FROM_MAP(mx, std::mutex, mutexes);
+    mx->lock();
+}
+
+bool DistributedCoordination::localTryLock(int32_t appId)
+{
+    FROM_MAP(mx, std::mutex, mutexes);
+    return mx->try_lock();
+}
+
+void DistributedCoordination::lock(const faabric::Message& msg)
+{
+    DISTRIBUTED_SYNC_OP(localLock, client.coordinationLock)
+}
+
+void DistributedCoordination::localUnlockRecursive(int32_t appId)
+{
+    FROM_MAP(mx, std::recursive_mutex, recursiveMutexes);
+    mx->unlock();
+}
+
+void DistributedCoordination::localUnlock(int32_t appId)
+{
+    FROM_MAP(mx, std::mutex, mutexes);
+    mx->unlock();
+}
+
+void DistributedCoordination::unlock(const faabric::Message& msg)
+{
+    DISTRIBUTED_SYNC_OP(localUnlock, client.coordinationUnlock)
+}
+
+void DistributedCoordination::doLocalNotify(int32_t appId, bool master)
+{
+    checkAppSizeSet(appId);
+
+    // All members must lock when entering this function
+    FROM_MAP(nowaitMutex, std::mutex, mutexes)
+    std::unique_lock<std::mutex> lock(*nowaitMutex);
+
+    FROM_MAP(nowaitCount, std::atomic<int>, counts)
+    FROM_MAP(nowaitCv, std::condition_variable, cvs)
+
+    int appSize = appSizes[appId];
+
+    if (master) {
+        auto timePoint = std::chrono::system_clock::now() +
+                         std::chrono::milliseconds(GROUP_TIMEOUT_MS);
+
+        if (!nowaitCv->wait_until(lock, timePoint, [&] {
+                return nowaitCount->load() >= appSize - 1;
+            })) {
+
+            SPDLOG_ERROR("Group {} await notify timed out", appId);
+            throw std::runtime_error("Group notify timed out");
+        }
+
+        // Reset, after we've finished
+        nowaitCount->store(0);
+    } else {
+        // If this is the last non-master member, notify
+        int countBefore = nowaitCount->fetch_add(1);
+        if (countBefore == appSize - 2) {
+            nowaitCv->notify_one();
+        } else if (countBefore > appSize - 2) {
+            SPDLOG_ERROR("Group {} notify exceeded group size, {} > {}",
+                         appId,
+                         countBefore,
+                         appSize - 2);
+            throw std::runtime_error("Group notify exceeded size");
+        }
+    }
+}
+
+void DistributedCoordination::localNotify(int32_t appId)
+{
+    doLocalNotify(appId, false);
+}
+
+void DistributedCoordination::awaitNotify(int32_t appId)
+{
+    doLocalNotify(appId, true);
+}
+
+void DistributedCoordination::notify(const faabric::Message& msg)
+{
+    DISTRIBUTED_SYNC_OP(localNotify, client.coordinationNotify)
+}
+
+void DistributedCoordination::localBarrier(int32_t appId)
```

Comment:
I am wondering if it is a good idea to add timeouts in this operations, to prevent applications waiting on barriers forever.

This has caused issues in the past, so maybe we could design the distributed coordination with this in mind.

Reply:
Yes, there are already timeouts implemented in the barriers (explicitly in this class or via the `faabric::util::Barrier` class), but I'll add them to the lock operations.